inline_topic_edit: Improve design of inline topic edit UI.

sayamsamal opened the pull request:
This PR improves the design of the inline topic edit UI by making the following changes:
- Increases the dimensions of the text input element.
- Fixes the font scaling inside text input element.
- Replaces the old save and cancel buttons with the new icon buttons.
- Hides all remaining recipient controls when the topic edit UI is displayed.

This PR also contains a commit adding an option in template to toggle squared icon buttons.

Fixes: #33155.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
| Before | After |
|--------|--------|
| ![Screenshot 2025-01-30 at 5 33 57 PM](https://github.com/user-attachments/assets/d6e1e689-2321-43eb-b19c-52c9fa21d343) | ![Screenshot 2025-01-30 at 5 34 16 PM](https://github.com/user-attachments/assets/78df0ba5-5636-4d35-a1e6-cca371176edf) | 
| ![Screenshot 2025-01-30 at 5 31 41 PM](https://github.com/user-attachments/assets/2b3142f2-e731-402f-be55-1924953ab645) | ![Screenshot 2025-01-30 at 5 29 55 PM](https://github.com/user-attachments/assets/16466506-90e7-4250-a564-9f4788395340) |

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

sayamsamal commented:
@alya could you give this PR a look?

sayamsamal commented:
The keyboard navigation for the inline topic edit UI seemed okay to me. Could you please check if you are facing the same keyboard navigation issue as you mentioned in the parent issue?

![Kapture 2025-01-30 at 23 01 20](https://github.com/user-attachments/assets/9d84f145-cc69-4e5e-88f9-d6855d8aefe9)

alya commented:
I tested the PR and posted some design thoughts for discussion [here](https://chat.zulip.org/#narrow/channel/101-design/topic/Improve.20design.20of.20topic.20edit.20UI.20in.20message.20header.20bar.20.2333155/near/2070402). In the meantime:

1. I'm not seeing the hover effect on the button that's in your screen capture. The icon changes color, but that's it.
2. If we stick with the icon buttons, we'll definitely need tooltips ("Save" and "Cancel" seem fine).

sayamsamal commented:
## Updates
- Fixed the missing hover effect on the buttons
- Added the "Save" and "Cancel" tooltips for the icon buttons.
- Added a max-width of 65ch to the inline topic input element.

| Light Mode | Dark Mode |
|--------|--------|
| ![Screenshot 2025-02-11 at 2 38 11 AM](https://github.com/user-attachments/assets/62a139f4-c520-4ec2-9839-a0f043a0f516) | ![Screenshot 2025-02-11 at 2 39 10 AM](https://github.com/user-attachments/assets/cd6e99cd-3d4a-40a3-b0b3-56a6df080363) |

alya commented:
Cool, to me this feels like a big improvement over the current UI!

alya commented:
As I'm playing with this, I'm getting into a state where renaming the topic doesn't navigate me to the new topic view. It's happening a lot, but I don't have a reproducer. Do you think it's related to the changes here?

![Screenshot 2025-02-10 at 23 27 05@2x](https://github.com/user-attachments/assets/2fb877e8-2701-4b7f-8f7a-88da09bf9759)

timabbott commented on web/templates/topic_edit_form.hbs:
```diff
@@ -1,10 +1,9 @@
 {{! Client-side Handlebars template for rendering the topic edit form. }}
 
 <form id="topic_edit_form">
```
Hmm, this should be a class, since there can be more than one copy of this element in the DOM at once.

sayamsamal replied:
Makes sense, updated.

sayamsamal commented:
> As I'm playing with this, I'm getting into a state where renaming the topic doesn't navigate me to the new topic view. It's happening a lot, but I don't have a specific reproducer. Do you think it's related to the changes here?
> 
> ![Screenshot 2025-02-10 at 23 27 05@2x](https://private-user-images.githubusercontent.com/2090066/411868683-2fb877e8-2701-4b7f-8f7a-88da09bf9759.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk2NjY3NjksIm5iZiI6MTczOTY2NjQ2OSwicGF0aCI6Ii8yMDkwMDY2LzQxMTg2ODY4My0yZmI4NzdlOC0yNzAxLTRiN2YtOGY3YS04OGRhMDliZjk3NTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMTZUMDA0MTA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZWM0ODE3ZTM1MTgyYmM3MTJkNTU4MzFhM2Y2M2EzYTU5MWZmYzk5ODY1OTY4NzVkOTcwMjBjMTFkNDUyZjY4ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.0MlgXTe0bzr8CapG5nuRjW51NfHp4vuiQ8YuIEVav7o)

This might have been due to the bug with the button defaulting to the "Submit" type and resulting in the submission of the form anytime Enter key is pressed in the inline topic edit UI. Could you please check if this has been fixed now?

sayamsamal commented:
## Updates
- Set the button type to "button" to prevent accidental submission of the form via the enter key and the issues with the keyboard handling logic in `handle_inline_topic_edit_keydown`.
-  Fixed the keyboard logic inside `handle_inline_topic_edit_keydown`, to fix bug where pressing Enter on the cancel button would also lead to the new topic name being set.
- Changed topic_edit_form from ID -> classname since multiple instances of the topic_edit_form can be used in a single DOM.
- Removed the overflowing error banner shown when the topic name is empty.
- Added red glow borders to the inline topic edit input element to indicate that the new topic name is being considered as empty according to the `realm_mandatory_topics property`
- Added methods to disable elements, and wrap them in a `span` wrapper t.o accommodate the tippy tooltips which could not have been triggered directly from the disabled elements since they don't fire events.

alya commented:
> Removed the overflowing error banner shown when the topic name is empty.

Could you please add a screenshot?

alya commented:
@karlstolley There's some ongoing discussion about the exact design of the buttons, but probably worth you taking a look at this point?

sayamsamal commented:
@alya I have added the screenshots/screencasts, showcasing the updates.

karlstolley commented on web/src/message_list.ts:
```diff
@@ -539,22 +539,17 @@ export class MessageList {
-        $recipient_row.find(".always_visible_topic_edit").hide();
-        $recipient_row.find(".on_hover_topic_resolve").hide();
-        $recipient_row.find(".on_hover_topic_unresolve").hide();
+        $recipient_row.find(".recipient_bar_control").hide();
```
I'd prefer going with our preferred hyphenated classes here, since this is being introduced anew.

sayamsamal replied:
Fixed.

karlstolley commented on web/styles/zulip.css:
```diff
@@ -858,38 +857,38 @@ div.focused-message-list.is-conversation-view .recipient_row {
+    /* Set the max width of the inline topic edit input to the realm's
+       max_topic_length (60 chars) + 5 chars extra for some cushion. */
+    max-width: 65ch;
+    line-height: 1.2142em;
```
This seems like a very specific number. How was it arrived at?

sayamsamal replied:
I have used the same `line-height` as that used for the stream topic:
```css
.message_header.message_header_stream .stream_topic {
    /* ... */
    line-height: 1.2142em;
    /* ... */
}
```

sayamsamal replied:
The logic behind this being that there should be minimal visual change when switching to the inline topic edit UI, giving it the feel that one is actually editing the text which was previously visible.

karlstolley commented:
Apart from a couple of nits, this looks good to me. It might be worth @prakhar1144 taking a quick look at the logic around mandatory topics/general chat, though.

alya commented:
The red outline in your screencapture seems to be missing the bottom part?

sayamsamal commented:
> The red outline in your screencapture seems to be missing the bottom part?

That would be an issue with the video recording, it is clearly visible in the UI:
![Screenshot 2025-02-21 at 8 55 58 PM](https://github.com/user-attachments/assets/a041d7f7-f3bf-4414-b636-069bb65fa312)

alya commented:
@timabbott This seems good enough to me, and an improvement on the existing UI. See [#design > Topic edit UI in message header bar #33155 / PR #33266](https://chat.zulip.org/#narrow/channel/101-design/topic/Topic.20edit.20UI.20in.20message.20header.20bar.20.2333155.20.2F.20PR.20.2333266/with/2105426).

timabbott commented on web/templates/components/icon_button.hbs:
```diff
@@ -1,3 +1,3 @@
-<button class="{{#if custom_classes}}{{custom_classes}} {{/if}}icon-button {{#if squared}}icon-button-square {{/if}}icon-button-{{intent}}" {{#if data-tooltip-template-id}}data-tooltip-template-id="{{data-tooltip-template-id}}"{{/if}} tabindex="0">
+<button type="button" class="{{#if custom_classes}}{{custom_classes}} {{/if}}icon-button {{#if squared}}icon-button-square {{/if}}icon-button-{{intent}}" {{#if data-tooltip-template-id}}data-tooltip-template-id="{{data-tooltip-template-id}}"{{/if}} tabindex="0">
```
Probably rather than these very long lines, we should consider doing some line-wrapping in these components and the handlebars partials that include them, to put one field per line or something.

sayamsamal replied:
Sure, I will work on this as a follow-up.

timabbott commented on web/src/ui_util.ts:
```diff
@@ -258,3 +259,27 @@ export function show_left_sidebar_menu_icon(element: Element): void {
+    const $tooltip_target_wrapper = $("<span>");
+    $tooltip_target_wrapper.addClass("disabled-tooltip");
+    $tooltip_target_wrapper.attr("data-tippy-content", tooltip_text).attr("tabindex", "0");
+    $element.wrap($tooltip_target_wrapper);
```
I don't see how this won't wrap an additional layer every time this is called.

timabbott replied:
Yep, confirmed.

![image](https://github.com/user-attachments/assets/1019cf7a-a147-46fa-b40f-95933d59f231)

timabbott replied:
Fixed like so:
``` diff
diff --git a/web/src/ui_util.ts b/web/src/ui_util.ts
index 485ffdb0f2..ae1d4e5757 100644
--- a/web/src/ui_util.ts
+++ b/web/src/ui_util.ts
@@ -278,6 +278,10 @@ export function disable_element_and_add_tooltip($element: JQuery, tooltip_text:
     // tippy tooltips on disabled elements. So, as a workaround, we wrap the
     // disabled element in a span and show the tooltip on this wrapper instead.
     // https://atomiks.github.io/tippyjs/v6/constructor/#disabled-elements
+    if ($element.prop("disabled")) {
+        // If already disabled, there's nothing to do.
+        return;
+    }
     $element.prop("disabled", true);
     const $tooltip_target_wrapper = $("<span>");
     $tooltip_target_wrapper.addClass("disabled-tooltip");
```

sayamsamal replied:
I understand, did not think of the corner case. Thank you for the fix!
